core/utils/syscallthread.h

```diff
@@ -0,0 +1,443 @@
+ * into a sub-object that is given to the thread, so that we can
+ * use thread_.detach() and knock_thread_.detach()).  If we do this
+ * we need new/dispose, std::unique_ptr<...>, and/or refcounts.
+ */
+template <bool ReliableSignals>
+class SyscallThread {
+ public:
+  // NB: order matters here.  Thread state progresses linearly
+  // (except for Reset() which must not be allowed to race).
+  enum ThreadState { TSNotStarted, TSStarting, TSReady, TSExiting, TSDone };
+
+  // to allow a future Detach, we have a WaitType enumeration
+  enum WaitType { RequestOnly, Wait };
+
+  SyscallThread()
+      : state_(TSNotStarted),
+        exit_requested_(false),
+        thread_(),
+        knock_thread_(),
+        defer_count_(0) {}
+
+  /*!
+   * Re-set state to allow re-firing thread.  USE WITH CAUTION!
+   * Returns false if you called it inappropriately, true if it
+   * did the reset.
+   *
+   * (If we ever allow detaching threads, this code will be
+   * responsible for allocating new state objects if needed.)
+   */
+  bool Reset() {
+    if (state_ == TSNotStarted) {
+      // Nothing to do.
+      return true;
+    }
+
+    if (state_ != TSDone) {
+      // Inappropriate call.
+      return false;
+    }
+
+    // Collect previous threads, if we have not yet.
+    WaitFor();
+
+    // Rewind state.
+    exit_requested_ = false;
+    state_ = TSNotStarted;
+    if (ReliableSignals) {
+      defer_count_ = 0;
+    }
+    return true;
+  }
+
+  /*!
+   * Start the thread running.  It will call the user provided Run()
+   * once it's ready (unless the thread was already asked to exit).
+   *
+   * Returns true on success, false (with errno set) on failure.
+   */
+  bool Start() {
+    // Establish process-wide signal catching, if that isn't
+    // yet done.  The signal handler will interrupt system calls.
+    if (!ExitSigMask::signals_initialized_) {
+      if (!CatchExitSignal()) {
+        return false;
+      }
+    }
+
+    // If the thread isn't in pristine state, this is an error.
+    if (state_ != TSNotStarted) {
+      errno = EINVAL;
+      return false;
+    }
+
+    assert(!thread_.joinable());
+    assert(!knock_thread_.joinable());
+
+    state_ = TSStarting;
+    thread_ = std::thread([this]() {
+      // Block all signals except (maybe) SIG_THREAD_EXIT.
+      sigset_t sigset;
+      sigfillset(&sigset);
+      if (ReliableSignals) {
+        // We're a reliable-signal case, so block SIG_THREAD_EXIT
+        // too.  The user promises to  be calling pselect/ppoll
+        // and unblocking it only for the duration of that call.
+        // This means that if it occurs in any other code path
+        // in the user's Run(), it will be deferred until the
+        // point at which the user calls pselect/ppoll.
+      } else {
+        // The user promises to call IsExitRequested() after
+        // each blocking system call.
+        sigdelset(&sigset, ExitSigMask::SIG_THREAD_EXIT);
+      }
+      ThreadSetSigmask(&sigset);
+
```

Comment:
May `ExitSigMask::allmask`, `ExitSigMask::sigmask` be used here for `sigset` as well?

Reply:
I think so, this was what's left after I consolidated some earlier versions that were more scattered (which is also why the class is in the header and not in the `.cc` file).